### PR TITLE
Add orders resource

### PR DESCRIPTION
**Pull request by enricostano**
More info: http://community.coopdevs.org/t/roadmap-to-fesc2016/70/3

To create an order **from a user to a group**:
```
POST /api/v1/orders
{
  to_group_id: integer,
  confirm_before: UTC timestamp
}
```
To create an order **from a group to a producer**:
```
POST /api/v1/orders
{
  from_group_id: integer,
  to_producer_id: integer,
  confirm_before: UTC timestamp
}
```

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -105,14 +105,14 @@
...
+    t.integer  "user_id",        null: false
+    t.integer  "group_id",       null: false
+    t.datetime "confirm_before", null: false
     t.datetime "created_at"
     t.datetime "updated_at"
   end
 
-  add_index "orders", ["group_id"], name: "index_orders_on_group_id", using: :btree
-  add_index "orders", ["user_id"], name: "index_orders_on_user_id", using: :btree
+  add_index "orders", ["user_id", "group_id", "confirm_before"], name: "index_orders_on_user_id_and_group_id_and_confirm_before", using: :btree
```
Comment (sauloperez) [inline db/schema.rb]: I don't see these there attributes used in any query. If so, I'd add only the id columns in the index

File: engines/suppliers/db/migrate/20150825155330_create_order.rb [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
+      t.references :user, null: false, foreign_key: true
+      t.references :group, null: false, foreign_key: true
+      t.datetime :confirm_before, null: false
 
       t.timestamps
     end
 
-    add_index :orders, [:user_id]
-    add_index :orders, [:group_id]
+    add_index :orders, [:user_id, :group_id, :confirm_before]
```
Comment (sauloperez) [inline engines/suppliers/db/migrate/20150825155330_create_order.rb]: Ditto

File: db/schema.rb (hunk 1 above)
Comment (enricostano) [inline db/schema.rb]: Wait for it... :trollface:

File: engines/suppliers/app/models/suppliers/order.rb [hunk 3]
```diff
@@ -1,10 +1,36 @@
...
+    validate :actor_presence
+    validate :target_presence
+    validates :confirm_before, presence: true
+
+    private
+
+    # Validates `from_user_id` xor `from_group_id` presence
+    #
+    def actor_presence
+      return if from_user_id.blank? ^ from_group_id.blank?
```
Comment (andresgutgon) [inline engines/suppliers/app/models/suppliers/order.rb]: What means '^'?

File: engines/suppliers/db/migrate/20150825155330_create_order.rb [hunk 4]
```diff
@@ -1,13 +1,35 @@
...
+            (from_group_id IS NOT NULL)::integer
+          ) = 1
+        ),
+        CHECK(
+          (
+            (to_group_id IS NOT NULL)::integer +
+            (to_producer_id IS NOT NULL)::integer
+          ) = 1
+        )
+      );
```
Comment (andresgutgon) [inline engines/suppliers/db/migrate/20150825155330_create_order.rb]: Why this SQL and not standard Rails migration?

Comment (sauloperez) [inline engines/suppliers/db/migrate/20150825155330_create_order.rb]: My only concern with this is that this things can't be saved in the db/schema as rails doesn't understand it. I believe the db/structure.sql either but I can't tell 100%.

File: engines/suppliers/app/models/suppliers/order.rb (hunk 3 above)
Comment (enricostano) [inline engines/suppliers/app/models/suppliers/order.rb]: it's an [exclusive OR](https://en.wikipedia.org/wiki/Exclusive_or): `a ^ b` means that `a` or `b` must exists but not both, just one of them.

File: engines/suppliers/db/migrate/20150825155330_create_order.rb (hunk 4 above)
Comment (enricostano) [inline engines/suppliers/db/migrate/20150825155330_create_order.rb]: @andresgutgon  Rails doesn't not support exclusive arcs, that's a better way to deal with polymorphism. Wherever I use it I put a link to more info in the migration. Read the doc on top of the file.

@sauloperez that's right, Rails doesn't keep track of custom queries. But that's not a concern IMHO. Rails doesn't offer a better way to deal with polymorphism. We cannot let Rails limit ourselves. There is no harm in using custom migrations, as usual you just need to know what you're doing.

Comment (sauloperez) [inline engines/suppliers/db/migrate/20150825155330_create_order.rb]: Yes, but we need to understand that our db/schema.rb won't be in sync and that can be troublesome. We better find out where/what relies on it.

**Comment by enricostano** [comment]
😬

**Pull request merged**